File: I3AutomationSite/ferramentas/previa_braco.py
# -*- coding: utf-8 -*-
"""Desenha a previa do braco a partir do JSON de previa_braco.js.

Rasteriza com o mesmo par de cores do site (linha #9FBEE0 sobre o navy do
heroi #001D3D) para que o que se ve aqui seja o que o navegador vai mostrar.

Uso: python ferramentas/previa_braco.py [giro] [inclinacao] [fase] [saida.png]
"""
import json
import subprocess
import sys
import os

from PIL import Image, ImageDraw

RAIZ = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
FUNDO = (0, 29, 61)
LINHA = (159, 190, 224)
ACESO = (253, 197, 0)
ESCALA = 2          # supersampling: linha de 1 px nao se julga em tela cheia


def render(giro=2.40, incl=0.16, fase=0.22, saida=None, larg=430, alt=584,
           dist=0, alvo_y=0, calado=False, ponteiro=None, corte=None):
    """larg/alt sao a moldura REAL do heroi: min(34vw, 430px) em 729/991."""
    js = subprocess.run(
        ["node", os.path.join(RAIZ, "ferramentas", "previa_braco.js"),
         str(giro), str(incl), str(fase), str(larg), str(alt),
         str(dist), str(alvo_y)],
        capture_output=True, text=True, cwd=RAIZ)
    if js.returncode != 0:
        print(js.stderr)
        raise SystemExit(js.returncode)
    d = json.loads(js.stdout)

    # Enquadramento medido, nao estimado: a caixa do desenho projetado contra
    # a caixa do slot. Sobra grande em cima e a peca "flutua"; negativo e corte.
    xs = [v for s in d["segs"] if s[4] > .05 for v in (s[0], s[2])]
    ys = [v for s in d["segs"] if s[4] > .05 for v in (s[1], s[3])]
    cx0, cx1, cy0, cy1 = min(xs), max(xs), min(ys), max(ys)
    if not calado:
        print("  caixa do desenho  x %.0f..%.0f de %d   y %.0f..%.0f de %d"
              % (cx0, cx1, d["larg"], cy0, cy1, d["alt"]))
        print("  preenchimento     %.0f%% da largura, %.0f%% da altura   "
              "distancia %.1f" % (100 * (cx1 - cx0) / d["larg"],
                                  100 * (cy1 - cy0) / d["alt"], d["dist"]))
        print("  folga             esq %.0f  dir %.0f  topo %.0f  base %.0f"
              % (cx0, d["larg"] - cx1, cy0, d["alt"] - cy1))

    W, H = int(d["larg"]) * ESCALA, int(d["alt"]) * ESCALA
    im = Image.new("RGB", (W, H), FUNDO)
    dr = ImageDraw.Draw(im, "RGBA")
    # `ponteiro` reproduz a lente do fragment shader por SEGMENTO (o shader faz
    # por fragmento). Serve para julgar a leitura do dourado, nao para medir.
    # `corte` reproduz a troca de folha do shader: abaixo dele a peca esta sobre
    # a secao clara, e a tinta vira navy com ganho maior. Em fracao da ALTURA,
    # medida do topo do canvas -- igual ao que medirCorte() calcula.
    if corte is not None:
        dr.rectangle((0, corte * H, W, H), fill=(255, 255, 255))
        dr.line((0, corte * H, W, corte * H), fill=(0, 0, 0, 36))
    for x0, y0, x1, y1, a, larg in d["segs"]:
        cor, ganho = LINHA, 1.0
        if corte is not None and (y0 + y1) / 2 > corte * d["alt"]:
            cor, ganho = (0, 53, 102), 1.15
        if ponteiro:
            px, py = ponteiro[0] * d["larg"], ponteiro[1] * d["alt"]
            raio = 0.30 * d["alt"]
            mx, my = (x0 + x1) / 2, (y0 + y1) / 2
            r = ((mx - px) ** 2 + (my - py) ** 2) ** .5 / raio
            rev = max(0.0, min(1.0, 1 - (r - .80) / .24))
            rev = rev * rev * (3 - 2 * rev)
            cor = tuple(int(LINHA[i] + (ACESO[i] - LINHA[i]) * rev) for i in range(3))
            ganho = 1 + rev * .42
            t = (r - .94) * 20
            banda = 2.718 ** (-t * t)
            if banda > .35:
                cor = ACESO
                ganho = max(ganho, 1.2)
        al = max(0, min(255, int(a * ganho * 255)))
        if al <= 2:
            continue
        dr.line((x0 * ESCALA, y0 * ESCALA, x1 * ESCALA, y1 * ESCALA),
                fill=cor + (al,), width=max(1, round(larg * ESCALA)))
    im = im.resize((int(d["larg"]), int(d["alt"])), Image.LANCZOS)
    saida = saida or os.path.join(RAIZ, "ferramentas", "_previa_braco.png")
    im.save(saida)
    print("%d segmentos, raio %.1f -> %s" % (d["segmentos"], d["raio"], saida))
    return saida


if __name__ == "__main__":
    a = sys.argv[1:]
    render(float(a[0]) if len(a) > 0 else 2.40,
           float(a[1]) if len(a) > 1 else 0.16,
           float(a[2]) if len(a) > 2 else 0.22,
           a[3] if len(a) > 3 else None)

File: I3AutomationSite/ferramentas/previa_braco.js
/* Prévia do braço FORA do navegador.

   Por que existe: a geometria de js/braco.js é gerada por código, e a única
   pergunta que importa — "isto se parece com um braço de 6 eixos?" — não se
   responde lendo o arquivo. Este harness carrega o MESMO módulo, roda a MESMA
   cinemática e a MESMA matriz MVP, e cospe os segmentos já projetados em JSON;
   ferramentas/previa_braco.py desenha o PNG.

   Não duplica nada: o código de produção é lido do disco e avaliado. Se a
   geometria mudar, a prévia muda junto — que é o ponto.

   Uso:  node ferramentas/previa_braco.js <giro> <inclinacao> <fase> > saida.json
*/
"use strict";
const fs = require("fs");
const path = require("path");
const vm = require("vm");

const ARQ = path.join(__dirname, "..", "site", "js", "braco.js");
/* gl.js tem de ser avaliado ANTES: desde a extracao do nucleo, braco.js pega
   matriz, parser de token e compilacao de shader de globalThis.I3GL. No
   navegador quem garante a ordem e o `defer` na ordem do documento; aqui e
   esta concatenacao. Sem ela o modulo lanca no primeiro alias e a previa
   morre calada. */
const NUCLEO = fs.readFileSync(
  path.join(__dirname, "..", "site", "js", "gl.js"), "utf8");
let fonte = fs.readFileSync(ARQ, "utf8");

/* Expõe os internos SEM tocar no arquivo de produção: o módulo é uma IIFE,
   então basta injetar a exportação imediatamente antes do fecho dela. */
const FECHO = "})();";
const i = fonte.lastIndexOf(FECHO);
if (i < 0) throw new Error("não achei o fecho da IIFE em braco.js");
fonte = fonte.slice(0, i) +
  "globalThis.__braco = { construir, mult, transladar, giroY, giroZ, " +
  "perspectiva, olhar, PIVO, PIVO_DEDO, enquadrar, MIRA, FOV, pose, CICLO };\n" +
  fonte.slice(i);

/* Stubs mínimos: o módulo só toca em matchMedia e no DOM depois de iniciar,
   e iniciar() não acha nenhum [data-braco] aqui. */
const contexto = {
  console,
  globalThis: null,
  window: {
    matchMedia: () => ({ matches: false, addEventListener() {} }),
    requestAnimationFrame() {},
    WebGL2RenderingContext: null
  },
  document: {
    readyState: "complete",
    querySelectorAll: () => [],
    addEventListener() {}
  }
};
contexto.globalThis = contexto;
vm.createContext(contexto);
vm.runInContext(NUCLEO, contexto);   /* publica contexto.I3GL */
vm.runInContext(fonte, contexto);
const B = contexto.__braco;

/* ---- mesma pose e mesma câmera do quadro real ---- */
const TAU = Math.PI * 2;
const giro = parseFloat(process.argv[2] ?? "2.40");
const incl = parseFloat(process.argv[3] ?? "0.16");
const fase = parseFloat(process.argv[4] ?? "0.22");
const LARG = parseFloat(process.argv[5] ?? "430");
const ALT = parseFloat(process.argv[6] ?? "584");
/* distancia e altura do alvo entram por argv para permitir varredura de
   enquadramento sem editar o arquivo de producao; os padroes sao os valores
   que estao em braco.js, entao rodar sem argumento mede o que o site faz. */
const DIST = parseFloat(process.argv[7] ?? "0") || null;
const ALVO_Y = parseFloat(process.argv[8] ?? "0") || null;

/* pose() NAO e reimplementada aqui: e a de site/js/braco.js, exposta acima.
   A copia que existia neste arquivo ja tinha divergido da original em uma
   sessao -- previa que diverge do original mede um site que nao existe.
   `fase` entra como fracao de 0..1 e vira tempo pelo ciclo real. */
const pose = (f) => B.pose(f * B.CICLO);

const cena = B.construir();
const dados = cena.segmentos;
const M = pose(fase);

/* A distancia sai da MESMA funcao que o site usa. DIST so existe para
   comparar contra um valor fixo; sem ele, isto mede o que o navegador faz. */
const mira = ALVO_Y || B.MIRA;
/* mesmo criterio do site: o maior de SEIS fases, para a distancia depender
   so do angulo e a peca nao pulsar de tamanho durante o ciclo */
let dAuto = 0;
for (let k = 0; k < 6; k++) {
  dAuto = Math.max(dAuto, B.enquadrar(cena.cantos, pose(k / 6), giro, incl, LARG / ALT));
}
const d = DIST || dAuto;
const olho = [
  Math.sin(giro) * Math.cos(incl) * d,
  mira + Math.sin(incl) * d,
  Math.cos(giro) * Math.cos(incl) * d
];
const mvp = B.mult(B.perspectiva(B.FOV, LARG / ALT, 1, 120),
                   B.olhar(olho, [0, mira, 0], [0, 1, 0]));

function aplicar(M, v) {
  const o = [0, 0, 0, 0];
  for (let r = 0; r < 4; r++) {
    o[r] = M[0 * 4 + r] * v[0] + M[1 * 4 + r] * v[1] +
           M[2 * 4 + r] * v[2] + M[3 * 4 + r] * v[3];
  }
  return o;
}

/* Raio da peca em torno do ponto visado -- mesma conta do site, e o que
   ancora a rampa de profundidade na distancia da camera. */
let RAIO = 1;
for (let e = 0; e < cena.cantos.length; e++) {
  const pts = cena.cantos[e], Me = M[e];
  if (!pts || !Me) continue;
  for (let i = 0; i < 8; i++) {
    const q = pts[i];
    const x = Me[0] * q[0] + Me[4] * q[1] + Me[8] * q[2] + Me[12];
    const y = Me[1] * q[0] + Me[5] * q[1] + Me[9] * q[2] + Me[13] - mira;
    const z = Me[2] * q[0] + Me[6] * q[1] + Me[10] * q[2] + Me[14];
    RAIO = Math.max(RAIO, Math.hypot(x, y, z));
  }
}

/* Larguras iguais as dos tokens do CSS (--braco-linha-fina/-grossa). */
const FINA = 0.9, GROSSA = 1.8;

const segs = [];
for (let k = 0; k < dados.length; k += 8) {
  const elo = Math.round(dados[k + 7]);
  const peso = dados[k + 6];
  const pontos = [];
  let prof = 1;
  for (const t of [[dados[k], dados[k + 1], dados[k + 2]],
                   [dados[k + 3], dados[k + 4], dados[k + 5]]]) {
    const mundo = aplicar(M[elo], [t[0], t[1], t[2], 1]);
    const p = aplicar(mvp, mundo);
    prof = Math.min(prof, Math.max(0.55, 1 - (p[3] - (d - RAIO)) / (2 * RAIO)));
    pontos.push([(p[0] / p[3] * 0.5 + 0.5) * LARG,
                 (1 - (p[1] / p[3] * 0.5 + 0.5)) * ALT]);
  }
  /* mesma curva do fragment shader, e a mesma largura do vertex shader */
  segs.push([pontos[0][0], pontos[0][1], pontos[1][0], pontos[1][1],
             (0.55 + 0.45 * peso) * 0.92 * prof,
             FINA + (GROSSA - FINA) * Math.max(0, Math.min(1, peso))]);
}
process.stdout.write(JSON.stringify({
  larg: LARG, alt: ALT, segmentos: segs.length, raio: RAIO,
  dist: d, segs
}));
